Actor manager

GoldSky17 opened the pull request:
Rename brach to scene manager and adds the ActorManager skeleton

Commit:
SceneManager

Commit:
Rename brach to scene manager and adds the ActorManager skeleton and logic for future integration

GoldSky17 committed:
Merge branch 'main' into actor_manager

Commit:
little comment update

Commit:
Merge branch 'actor_manager' of https://github.com/mefkek/cpp_game into actor_manager

Commit:
add Party.hpp, Actor, ActorBehaviour.hpp, and ActorRaceEnum.hpp

- extracted and implemented Actor, Party, ActorManager, ActorBehaviour, and ActorRaceEnum classes
- added update loop with automatic removal of dead actors

Commit:
change source path in Actor.cpp

mefkek commented on include/Nodes/Actor.hpp:
```diff
@@ -0,0 +1,36 @@
+
+class Actor : public Node {
+public:
+    Actor(ActorRaceEnum raceIn, std::unique_ptr<ActorBehaviour> behav);
```
behav can't be a unique_ptr as many actor's can share the same bahaviour

mefkek commented on include/Nodes/ActorBehaviour.hpp:
```diff
@@ -0,0 +1,10 @@
+public:
+    virtual ~ActorBehaviour() = default;
+
+    virtual void update(Actor& actor, float delta) = 0;
```
To avoid confusion with Node::update this should be renamed, also it does not need a delta

mefkek commented on include/Nodes/ActorManager.hpp:
```diff
@@ -0,0 +1,27 @@
+#include <vector>
+#include <memory>
+#include <algorithm> // std:remove_if
+// Forward declaration Actor class
```
This is not a forward declaration

mefkek commented on src/Nodes/Actor.cpp:
```diff
@@ -0,0 +1,20 @@
+}
+
+void Actor::changeStat(const std::string& name, int delta) {
+    int& val = stats[name];
```
add a check that throw's an exception when name isn't  a key

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,47 @@
+#include <Nodes/ActorManager.hpp>
+#include "Nodes/ActorRace.hpp"
+
+void ActorManager::addActor(const std::shared_ptr<Actor>& actor) {
```
should create and return an actor, not copy it

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,47 @@
+
+}
+
+void ActorManager::update(float delta) {
```
Actor is a node, it doesn't make sense to update it maunually. Instead it should be a child of ActorManager or a Party. Dead actor's removing logic should be handled in Actor::update.

mefkek commented on include/Nodes/Party.hpp:
```diff
@@ -0,0 +1,33 @@
+#include <algorithm>
+#include "Actor.hpp"
+
+// Grupa aktorow
```
Should be a node and a child of ActorManager

mefkek reviewed (changes requested):
- change code formatting to fit the standard
- resolve the issues mentioned in other comments
- when events branch will be merged to main this should be rewritten to use features such as custom events and ContainerNode

Commit:
Party now inherits from Node, holds an actor vector, and delegates its update() call

ActorManager creates Party and Actor via add_child<>, replacing manual vector management

Removed the = default constructor declaration from ActorManager’s header and moved its definition into the .cpp

Fixed compilation errors around accessing the protected parent and removing dead actors

mefkek commented on include/Nodes/ActorManager.hpp:
```diff
@@ -0,0 +1,28 @@
+#include "Nodes/ActorBehaviour.hpp"
+#include <memory>
+
+class ActorManager : public Node, public std::enable_shared_from_this<ActorManager> {
```
Node already inherits shared_from_this

mefkek replied:
Node already inherits shared_from_this

mefkek commented on include/Nodes/Node.hpp:
```diff
@@ -22,17 +22,19 @@ class Node : public std::enable_shared_from_this<Node>
     template <typename T, typename ... Args>
-    void add_child(Args&&... args)
+    //void add_child(Args&&... args)
+    std::shared_ptr<T> add_child(Args&&... args)
```
add_child already returns a pointer on main branch, merge to include

mefkek commented on include/Nodes/Party.hpp:
```diff
@@ -0,0 +1,46 @@
+#include <algorithm>
+
+// Grupa aktorów jako węzeł w drzewie sceny
+class Party : public Node {
```
Merge with main and revrite as ContainerNode<Actor>

mefkek replied:
When merged with main branch rewrite this class as a ContainerNode
```suggestion
class Party : public ContainerNode<Actor>
```
You should then be able to use the functions specific to ContainerNode instead

mefkek commented on include/Nodes/Party.hpp:
```diff
@@ -0,0 +1,46 @@
+    }
+
+private:
+    std::vector<std::shared_ptr<Actor>> actors;
```
actors should be stored as children

mefkek replied:
actors should be stored as children, and instanced as
`add_child<Actor>(Args some_arguments)`

mefkek commented on src/Nodes/Actor.cpp:
```diff
@@ -0,0 +1,27 @@
+    auto it = stats.find(name);
+    if (it == stats.end()) {
+        // rzucamy wyjątek, jeśli statystyka o podanej nazwie nie istnieje
+        throw std::out_of_range("Stat '" + name + "' does not exist");
```
I would make it a std::runtime_error, but isn't exactly wrong. This should propably be a warning though. If you start implementing events please aslo include a class that changed the stat into the warning message, something like
```suggestion
        Logger::log(Logger::MessageType::Warning, caller, 
                           " tried changing non existant stat ", name);
```

mefkek replied:
I would make it a std::runtime_error, but isn't exactly wrong. This should propably be a warning though. If you start implementing events please aslo include a class that changed the stat into the warning message, something like
`
        Logger::log(Logger::MessageType::Warning, caller, 
                           " tried changing non existant stat ", name);
`

mefkek commented on include/Nodes/ActorBehaviour.hpp:
```diff
@@ -0,0 +1,10 @@
+#pragma once
+class Actor;
```
Include a Actor.hpp instead

mefkek replied:
Include a Actor.hpp instead

mefkek commented on include/Nodes/Node.hpp:
```diff
@@ -22,17 +22,19 @@ class Node : public std::enable_shared_from_this<Node>
     std::vector<std::shared_ptr<Node>>& get_children();
 
     template <typename T, typename ... Args>
-    void add_child(Args&&... args)
```
add_child already returns a pointer on main branch, merge to include

mefkek commented on include/Nodes/Party.hpp:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    // Deleguje update na wszystkich aktorów w grupie
+    void update(float delta) override {
```
Actors and Party are both nodes, they get update as long as they are a child of a Node in a tree.
No need to call update manually

mefkek commented on include/Nodes/Actor.hpp:
```diff
@@ -0,0 +1,37 @@
+#include <unordered_map>
+#include <string>
+#include <memory>
+#include <stdexcept>    // dodane dla std::out_of_range
```
If this will be used it can be moved into Actor.cpp, no need for a comment

mefkek commented on src/Nodes/Actor.cpp:
```diff
@@ -0,0 +1,27 @@
+    // aktualizacja wartości
+    it->second += delta;
+    // wywołanie hooka po zmianie
+    onStatChange(name, it->second);
```
When merged with main rewrite as a CustomEvent, thouhght the arguments should also include the class that changed the stat to make debugging easier. Reference definition for details.

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,63 @@
+    return actor;
+}
+
+void ActorManager::update(float delta) {
```
As before, update is already called elsewere

mefkek reviewed (changes requested):
### TODO
- code formating is not compliant with the rest of the code base
- comments should be written in english
- resolve issues mentioned in comments

Commit:
Test

Commit:
Merge branch 'main' into actor_manager

# Conflicts:
#	include/Nodes.hpp
#	include/Nodes/Node.hpp
#	src/CMakeLists.txt

Commit:
Added an out-of-line definition for the virtual destructor in WindowEventManager.cpp to ensure the vtable is generated and resolve linking errors also restore the issue(If I've forgotten anything, please let me know.)

GoldSky17 committed:
Merge branch 'main' into actor_manager

mefkek committed:
Reverted bad changes

mefkek committed:
Merge branch 'reverting_bad_changes' into actor_manager

Removed bad changes, the errors in previous version are caused by improper merginng (i guess) and lack of code understanding. Class node has a pure virtual function update than needs(!!!) to be overwritten, and the Node::initialize() function exists
so there is an constructor like structure that works an an existing shared_ptr to avoid bad_weak_ptr exception. The only thing that can be done about is maybe adding a property_setter function that takes a pointer to NodeBuilder class object as an
argument or something like that.

mefkek commented on include/Nodes/ActorManager.hpp:
```diff
@@ -0,0 +1,25 @@
+#include <memory>
+
+// Manages creation and grouping of Actor instances.
+class ActorManager : public Node
```
This will generte an error becouse ActorManager need to override void update(float)

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+ActorManager::ActorManager()
+{
+    // Create the player party as a child of this manager
+    add_child<Party>();
```
This will generate bad_weak_ptr exception becouse it acts on an shared_ptr which doesn't exists yet (the object is beeing constructed). Move to overwritten initialize function.

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+        race == ActorRaceEnum::Lich;
+
+    // 3) Add the same actor into the appropriate party
+    auto targetParty = isEnemy ? enemyParty_ : playerParty_;
```
Actor can not be in a party

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // 2) Determine whether this actor is an enemy based on race
+    const bool isEnemy =
```
That is racist, use a separate argument for deciding where the actor will go, or overwrite the Party functions to automatically pass the actor to ActorManager

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+    // 3) Add the same actor into the appropriate party
+    auto targetParty = isEnemy ? enemyParty_ : playerParty_;
+    if (targetParty) {
+        targetParty->add_child<Actor>(race, behaviour);
```
Use ContainerNode::add_element instead

mefkek commented on include/Nodes/ActorBehaviour.hpp:
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "Actor.hpp"
+class  Actor;
```
You included Actor.hpp, this is not needed

mefkek commented on include/Nodes/ActorBehaviour.hpp:
```diff
@@ -0,0 +1,13 @@
+    virtual ~ActorBehaviour() = default;
+
+    // Called each frame to drive the actor’s logic.
+    virtual void behave(Actor& actor) = 0;
```
Use a weak_ptr instead of a reference for consistency

mefkek commented on src/Nodes/Actor.cpp:
```diff
@@ -0,0 +1,50 @@
+{
+    if (behaviour_)
+    {
+        behaviour_->behave(*this);
```
use shared_from_this() instead of a pointer

mefkek commented on src/Nodes/Actor.cpp:
```diff
@@ -0,0 +1,50 @@
+    {
+        // Log a warning if the stat does not exist, then invoke failure hook
+        const std::string caller = typeid(*this).name();
+        Logger::log(
```
replace caller with shared_from_this(), the overloaded operator<< in Node takes care of everything.

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+    std::shared_ptr<ActorBehaviour> behaviour)
+{
+    // 1) Instantiate the actor under this manager
+    add_child<Actor>(race, behaviour);
```
this already returns a weak_ptr, you can just call lock() on it

mefkek commented on src/Nodes/ActorManager.cpp:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+std::shared_ptr<Actor> ActorManager::addActor(
```
To avoid issues with object deleteion this should be a weak_ptr

mefkek commented on src/Nodes/Node.cpp:
```diff
@@ -52,4 +52,10 @@ std::ostream& operator<<(std::ostream& os, const std::shared_ptr<Node>& n)
\ No newline at end of file
+}
+
+std::ostream& operator<<(std::ostream& os, std::shared_ptr<Node>& n)
```
i accidentally missed it, please delete this as it will cause issues

mefkek reviewed (changes requested):
- Refer to my commit to see what i changed back
 - Most of your error stem from either a pure virtual function that you did not override, a shared_from_this() beeing used in constructor (which is the reason why initialize function exists, although it could be imporved. I'll make an issue about this), or accidentally deleteing a subfolder with a library that you need to link to from CMakeLists.txt

Commit:
Resolve the issue from feedback

mefkek reviewed (commented):
Generally good, add  a function for loading/writing to string/stringstream

Commit:
adding a function for loading\writing to string

mefkek reviewed (commented):
Should be good

Commit:
Merge branch 'refs/heads/main' into actor_manager

Commit:
Enhancing functionality and maintaining compatibility with existing code

GoldSky17 committed:
- Refactor Party::display to accept position, map scale and sprite scale; update actor positioning and sprite scaling
- Update Application to call Party::display with separate map and sprite scales
- Implement ActorManager::addParty for dynamic party creation
- Add RoomVisualizer::visualize() and move() methods to combine minimap and room rendering
- Extend display_minimap to choose tiles based on corridor exits and render player icon

GoldSky17 committed:
Merge branch 'main' into actor_manager

mefkek committed:
Added ActorCards

mefkek committed:
Merge branch 'main' into actor_manager to include new features

mefkek committed:
Made it so actors appear on session start and dissaper on session end

GoldSky17 committed:
Add ActorCard UI component and integrate into party display

- Introduce ActorCard class (ActorCard.hpp/.cpp) for rendering actor portrait, name and HP on the Debug_ui layer :contentReference[oaicite:0]{index=0}
- Extend Party::display_cards() to spawn an ActorCard for each party member :contentReference[oaicite:1]{index=1}
- Invoke display_cards() during Application initialization to render the cards alongside sprites :contentReference[oaicite:2]{index=2}

This change enables a visual “card” for each actor in the party, showing their portrait, name and current HP, and keeps the UI in sync by updating the HP label each frame.

mefkek committed:
Changed event that dungeon menager reacts to

mefkek committed:
Merge branch 'actor_manager' of https://github.com/mefkek/cpp_game into actor_manager
